### PR TITLE
[WFGP-222] Generate YAML channel definition

### PR DESCRIPTION
**Pull request by jmesnil**
When a feature-pack is built, create a new artifact (with "channel"
classifier" and "yaml" extension) that represents the channel definition
for the feature pack.
This feature is controlled by the generate-channel
parameter (optional, false  by default)

Dependencies to other feature-packs are defined in the channel's
"requires" field. They can be skipped if
add-feature-packs-as-required-channels is set to false (optional, true
by default).

JIRA: https://issues.redhat.com/browse/WFGP-222

Signed-off-by: Jeff Mesnil <jmesnil@redhat.com>

### CODE REVIEW

**Comment by jmesnil** [comment]
@jfdenise Could you please review?

**Comment by jfdenise** [comment]
I think that we should merge this new feature once we bumped major version to 6.

File: maven-plugin/src/main/java/org/wildfly/galleon/maven/ArtifactListBuilder.java [hunk 1]
```diff
@@ -16,6 +16,8 @@
  */
```
Comment (jfdenise) [inline maven-plugin/src/main/java/org/wildfly/galleon/maven/ArtifactListBuilder.java:16]: I don't think that this class needs to be updated.

File: maven-plugin/src/main/java/org/wildfly/galleon/maven/AbstractFeaturePackBuildMojo.java [hunk 2]
```diff
@@ -410,6 +434,14 @@ protected void buildFeaturePack(FeaturePackDescription.Builder fpBuilder, WildFl
                                 debug("Attaching feature-pack artifact list %s as a project artifact", offLinerTarget);
                                 Files.write(offLinerTarget, builder.build().getBytes());
                                 projectHelper.attachArtifact(project, ARTIFACT_LIST_EXTENSION, ARTIFACT_LIST_CLASSIFIER, offLinerTarget.toFile());
+                                final Path channelTarget = Paths.get(project.getBuild().getDirectory()).resolve(artifactId + '-'
```
[the remark this answers is not in the document]
Comment (jmesnil) [reply maven-plugin/src/main/java/org/wildfly/galleon/maven/AbstractFeaturePackBuildMojo.java]: fixed

File: maven-plugin/src/main/java/org/wildfly/galleon/maven/AbstractFeaturePackBuildMojo.java [hunk 3]
```diff
@@ -420,6 +452,45 @@ protected void buildFeaturePack(FeaturePackDescription.Builder fpBuilder, WildFl
...
+                }
+            }
+        }
+        // add a stream for this feature pack
+        channel.append("streams:\n");
+        appendMavenGAV(channel, "  ", project.getGroupId(),
+                project.getArtifactId(),
+                project.getVersion());
+        // add all feature-pack artifacts dependencies as streams
+        builder.appendJarArtifactsAsStreams(channel);
```
[the remark this answers is not in the document]
Comment (jmesnil) [reply maven-plugin/src/main/java/org/wildfly/galleon/maven/AbstractFeaturePackBuildMojo.java]: fixed.

I had to filter again the MavenProjectArtifactVersions.getFilteredArtifacts(project, buildConfig) to make sure I was only adding the `jar` dependencies in the component's streams.

File: maven-plugin/src/main/java/org/wildfly/galleon/maven/ArtifactListBuilder.java (hunk 1 above)
Comment (jmesnil) [reply maven-plugin/src/main/java/org/wildfly/galleon/maven/ArtifactListBuilder.java]: fixed

File: maven-plugin/src/main/java/org/wildfly/galleon/maven/AbstractFeaturePackBuildMojo.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jfdenise) [reply maven-plugin/src/main/java/org/wildfly/galleon/maven/AbstractFeaturePackBuildMojo.java]: We should perhaps ignore this comment and only handle jars but I am thinking at corner cases there, we could have a feature-pack that contains reference (from a copy-task) to other kind of artifacts (eg: a war) in maven.

[the remark this answers is not in the document]
Comment (jmesnil) [reply maven-plugin/src/main/java/org/wildfly/galleon/maven/AbstractFeaturePackBuildMojo.java]: mmh, it'd be better to ignore the "zip" dependencies that have been added as `required` channel and keep all other types. I'll update my PR

[the remark this answers is not in the document]
Comment (jmesnil) [reply maven-plugin/src/main/java/org/wildfly/galleon/maven/AbstractFeaturePackBuildMojo.java]: I've updated the code to ignore `zip` and `pom` artifacts

**Review by jfdenise — APPROVED** [review]

**Pull request merged**